docs/docs/tutorials/security/authorization.mdx

```diff
@@ -0,0 +1,1064 @@
+
+Prior to running the code below, ensure that you have a `.env` file with the following content:
+
+```env
+# Server host/port
+HOST=localhost
+PORT=3000
+
+# Auth server location
+AUTH_HOST=localhost
+AUTH_PORT=8080
+AUTH_REALM=master
+
+# Keycloak OAuth client credentials
+OAUTH_CLIENT_ID=<YOUR_SERVER_CLIENT_ID>
+OAUTH_CLIENT_SECRET=<YOUR_SERVER_CLIENT_SECRET>
+```
+
+`OAUTH_CLIENT_ID` and `OAUTH_CLIENT_SECRET` are associated with the MCP server client we created earlier.
+
+In addition to implementing the MCP authorization specification, the server below also does token introspection via Keycloak to make sure that the token it receives from the client is valid. It also implements basic logging to allow you to easily diagnose any issues.
+
+```typescript
+import 'dotenv/config';
+import express from "express";
+import { randomUUID } from "node:crypto";
+import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
+import { StreamableHTTPServerTransport } from "@modelcontextprotocol/sdk/server/streamableHttp.js";
+import { isInitializeRequest } from "@modelcontextprotocol/sdk/types.js";
+import { z } from "zod";
+import cors from "cors";
+import { mcpAuthMetadataRouter, getOAuthProtectedResourceMetadataUrl } from "@modelcontextprotocol/sdk/server/auth/router.js";
+import { requireBearerAuth } from "@modelcontextprotocol/sdk/server/auth/middleware/bearerAuth.js";
+import { OAuthMetadata } from "@modelcontextprotocol/sdk/shared/auth.js";
+import { checkResourceAllowed } from "@modelcontextprotocol/sdk/shared/auth-utils.js";
+const CONFIG = {
+  host: process.env.HOST || "localhost",
+  port: Number(process.env.PORT) || 3000,
+  auth: {
+    host: process.env.AUTH_HOST || process.env.HOST || "localhost",
+    port: Number(process.env.AUTH_PORT) || 8080,
+    realm: process.env.AUTH_REALM || "master",
+    clientId: process.env.OAUTH_CLIENT_ID || "mcp-server",
+    clientSecret: process.env.OAUTH_CLIENT_SECRET || "",
+  },
+};
+
+function createOAuthUrls() {
+  const authBaseUrl = new URL(`http://${CONFIG.auth.host}:${CONFIG.auth.port}/realms/${CONFIG.auth.realm}/`);
+  return {
+    issuer: authBaseUrl.toString(),
+    introspection_endpoint: new URL("protocol/openid-connect/token/introspect", authBaseUrl).toString(),
+    authorization_endpoint: new URL("protocol/openid-connect/auth", authBaseUrl).toString(),
+    token_endpoint: new URL("protocol/openid-connect/token", authBaseUrl).toString(),
+  };
+}
+
+function createRequestLogger() {
+  return (req: any, res: any, next: any) => {
+    const start = Date.now();
+    res.on('finish', () => {
+      const ms = Date.now() - start;
+      console.log(`${req.method} ${req.originalUrl} -> ${res.statusCode} ${ms}ms`);
+    });
+    next();
+  };
+}
+
+const app = express();
+
+app.use(express.json({
+  verify: (req: any, _res, buf) => {
+    req.rawBody = buf?.toString() ?? '';
+  }
+}));
+
+app.use(cors({
+  origin: '*',
```

Comment:
It may be worth pointing to the [spec](https://modelcontextprotocol.io/specification/2025-06-18/basic/transports#security-warning) security warning somewhere to clarify why we use CORS here, and e.g. add a comment that this is not production-ready. See also my [test](https://github.com/modelcontextprotocol/csharp-sdk/discussions/687#discussioncomment-14044988 ) using the c# sdk.

For development and guide-purposes, I think using a localhost-uri is a better default than `*`, wdyt?